crates/jstzd/src/task/task.rs

```diff
@@ -0,0 +1,253 @@
+use anyhow;
+use async_dropper::AsyncDrop;
+use tokio::task::JoinHandle as tokioJoinHandle;
+
+use async_trait::async_trait;
+use futures_util::{
+    task::{Context, Poll},
+    Future,
+};
+
+use std::pin::Pin;
+use tokio::signal::unix::{Signal, SignalKind};
+use tokio::sync::{mpsc, OnceCell};
+
+#[async_trait]
+pub trait Task {
+    /// A set of tasks which are all executed on the same 'execution backend'
+    type TaskSet;
+
+    /// Performs the unit of work for the task
+    async fn run_task(
+        self,
+        signals: mpsc::Receiver<Signal>,
+        task_group: &Self::TaskSet,
+    ) -> anyhow::Result<()>;
+
+    /// Spawns the task
+    async fn spawn_task(self) -> JoinHandle;
+}
+
+pub struct JoinHandle {
+    task: OnceCell<tokioJoinHandle<()>>,
+    allowed_signal: SignalKind,
+    dropped: bool,
+}
+
+// Need this because SignalKind does not have default
+impl Default for JoinHandle {
+    fn default() -> Self {
+        JoinHandle {
+            task: OnceCell::new(),
+            allowed_signal: SignalKind::terminate(),
+            dropped: false,
+        }
+    }
+}
+
+impl JoinHandle {
+    /// Creates a new join handle from a task.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use jstzd::task::task::JoinHandle;
+    /// use tokio::signal::unix::SignalKind;
+    ///
+    /// # tokio_test::block_on(async {
+    /// let handle = JoinHandle::new(
+    ///     tokio::spawn(async move {
+    ///         tokio::time::sleep(std::time::Duration::from_secs(10)).await;
+    ///     }),
+    ///     SignalKind::terminate(),
+    /// );
+    /// # });
+    /// ```
+    pub fn new(task: tokioJoinHandle<()>, signal: SignalKind) -> Self {
```

Comment:
Not really intended to be part of the API, recommend guarding with `#[cfg(test)]`